### PR TITLE
[MLDM-11] Add JobContext to PJS

### PR DESCRIPTION
**Pull request by FahadBSyed**

### CODE REVIEW

File: src/internal/pjsdb/job_context.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	if numRows == 0 {
+		return sql.ErrNoRows
+	}
+	return nil
+}
+
+func generateToken() []byte {
+	uuidv7, err := uuid.NewV7()
+	if err != nil {
+		panic(err) // the uuid generation can error if the random generator's mutex fails to unlock. Seems unlikely.
```
Comment (Zhang-Muyang) [inline src/internal/pjsdb/job_context.go:83]: A question: When should we return an error and when should we panic?

File: src/internal/pjsdb/job_context.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	if err != nil {
+		return errors.Wrap(err, "revoke job context")
+	}
+	if numRows == 0 {
+		return sql.ErrNoRows
+	}
+	return nil
+}
+
+func generateToken() []byte {
```
Comment (Zhang-Muyang) [inline src/internal/pjsdb/job_context.go:80]: return type is JobContextToken?

File: src/internal/pjsdb/job_context.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	"github.com/google/uuid"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachhash"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+)
+
+var (
+	//HashFn the hash function used by the pjsdb library for hashing job context tokens.
+	HashFn = hashFn
```
[the remark this answers is not in the document]
Comment (FahadBSyed) [reply src/internal/pjsdb/job_context.go:15]: Yes, for testing. You can swap it out.

File: src/internal/pjsdb/job_context.go (hunk 1 above)
Comment (FahadBSyed) [reply src/internal/pjsdb/job_context.go:83]: Panic is when we don't think we're really ever going to error. Error is for when we are sometimes expecting errors.

File: src/internal/pjsdb/job_context_test.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	t.Run("valid/exists", func(t *testing.T) {
+		withDependencies(t, func(d dependencies) {
+			id, err := createJob(t, d, 0)
+			require.NoError(t, err)
+			_, err = pjsdb.CreateJobContext(d.ctx, d.tx, id)
+			require.NoError(t, err)
+			require.NoError(t, pjsdb.RevokeJobContext(d.ctx, d.tx, id))
+			jobCtx := &pjsdb.JobContext{}
+			err = d.tx.GetContext(d.ctx, jobCtx, `SELECT context_hash FROM pjs.jobs WHERE id = $1 AND context_hash IS NOT NULL`, id)
+			t.Log(jobCtx, err)
```
Comment (Zhang-Muyang) [inline src/internal/pjsdb/job_context_test.go:97]: Should we keep the log?

File: src/internal/pjsdb/job_context.go (hunk 2 above)
Comment (FahadBSyed) [reply src/internal/pjsdb/job_context.go:80]: Good idea!

**Review by Zhang-Muyang — APPROVED** [review]

**Pull request merged**